app/upgrades_test.go

```diff
@@ -1022,3 +1023,236 @@
+		// So I do some math to make it seem kind of random, but is still deterministic.
+		// 7, 39, and 79 were picked randomly and have no special meaning.
+		// 4,999 was chosen so that there's 3,333 possible results.
+		// This will overflow at 2,097,112, but it shouldn't get bigger than 300,000 in here, so we ignore that.
+		j := ((i + 7) * (i + 39) * (i + 79)) % 49_999
+		return metadatatypes.ScopeSpecMetadataAddress(newUUID(j))
+	}
+	newScope := func(i int) metadatatypes.Scope {
+		rv := metadatatypes.Scope{
+			ScopeId:            newScopeID(i),
+			SpecificationId:    newSpecID(i),
+			RequirePartyRollup: i%6 > 2, // 50% chance, three false, then three true, repeated.
+		}
+
+		// 1 in 7 does not have a value owner.
+		incVO := i%7 != 0
+		if incVO {
+			rv.ValueOwnerAddress = addrs[i%len(addrs)]
+		}
+
+		// Include 1 to 5 owners and make one of them the value owner in just under 1 in 11 scopes.
+		ownerCount := (i % 5) + 1
+		incVOInOwners := incVO && i%11 == 0
+		incVOAt := i % ownerCount
+		rv.Owners = make([]metadatatypes.Party, ownerCount)
+		for o := range rv.Owners {
+			a := i + (o+1)*300
+			if incVOInOwners && o == incVOAt {
+				a = i
+			}
+			rv.Owners[o].Address = addrs[a%len(addrs)]
+			rv.Owners[o].Role = metadatatypes.PartyType(1 + (i+o)%11) // 11 different roles, 1 to 11.
+		}
+
+		daCount := (i % 3) // 0 to 2.
+		for d := 0; d < daCount; d++ {
+			a := i + (d+1)*33
+			rv.DataAccess = append(rv.DataAccess, addrs[a%len(addrs)])
+		}
+
+		return rv
+	}
+	voInOwners := func(scope metadatatypes.Scope) bool {
+		if len(scope.ValueOwnerAddress) == 0 {
+			return false
+		}
+		for _, owner := range scope.Owners {
+			if owner.Address == scope.ValueOwnerAddress {
+				return true
+			}
+		}
+		return false
+	}
+
+	// Create 300,005 scopes and write them to state.
+	// 6 of 7 have a value owner = 257_147 = 300_005 * 6 / 7 (truncated).
+	// 1 of 7 do not = 42_858 = 300_005 - 257_147 .
+	// There's 23,377 scopes that have the value owner in owners.
+	// So we expect 490,917 keys to delete = 257,147 * 2 - 23,377.
+	expCoin := make([]metadatatypes.MetadataAddress, 0, 257_147)
+	expNoCoin := make([]metadatatypes.MetadataAddress, 0, 42_858)
+	expDelInds := make([][]byte, 0, 490_917)
+	expBals := make(map[string]sdk.Coins)
+	t1 := time.Now()
+	for i := 0; i < 300_005; i++ {
+		scope := newScope(i)
+		s.Require().NoError(s.app.MetadataKeeper.V3WriteNewScope(s.ctx, scope), "[%d]: V3WriteNewScope", i)
+
+		if len(scope.ValueOwnerAddress) == 0 {
+			expNoCoin = append(expNoCoin, scope.ScopeId)
+			continue
+		}
+
+		expCoin = append(expCoin, scope.ScopeId)
+
+		vo := scope.ValueOwnerAddress
+		voAddr := sdk.MustAccAddressFromBech32(vo)
+		expDelInds = append(expDelInds, metadatakeeper.GetValueOwnerScopeCacheKey(voAddr, scope.ScopeId))
+		if !voInOwners(scope) {
+			expDelInds = append(expDelInds, metadatatypes.GetAddressScopeCacheKey(voAddr, scope.ScopeId))
+		}
+
+		expBals[vo] = expBals[vo].Add(scope.ScopeId.Coin())
+	}
```

Comment:
It creates 300,005 scopes to run the test on because when it runs live, it'll be running on about that same amount. So this unit test gives us a glimpse of what will happen.